cpu/atmega_common/periph/rtc.c

```diff
@@ -0,0 +1,218 @@
+
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    time_secs = (time_t)TCNT2;
+
+    /* Convert to seconds (highest 3 bits of TCNT2) */
+    time_secs = (time_secs & 0x000000E0) >> 5;
+
+    time_secs += rtc_state.time;
+
+    /* Convert from seconds since the epoch */
+    gmtime_r(&time_secs, time);
+
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2B);
+    rtc_state.alarm_cb = NULL;
+
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTC sleeps until safe to write OCR2B\n");
+    _asynch_wait();
+
+    /* Set alarm time */
+    rtc_state.alarm = mk_gmtime(time);
+
+    /* Prepare the counter for sub 8-second precision */
+    OCR2B = ((uint8_t)rtc_state.alarm & 0x07) << 5;
+
+    /* Interrupt safe order of assignment */
+    rtc_state.alarm_arg = arg;
+    rtc_state.alarm_cb = cb;
+
+    /* Enable alarm only if it will trigger in < 8 seconds */
+    if ((rtc_state.alarm & 0xFFFFFFF8) <= (rtc_state.time & 0xFFFFFFF8)) {
+        if (rtc_state.alarm <= rtc_state.time) {
+            /* Prevent alarm offset if time is too soon */
+            rtc_state.alarm_cb(rtc_state.alarm_arg);
+        }
+        else {
+            TIMSK2 |= (1 << OCIE2B);
+        }
+    }
+
+    return 0;
+}
+
+int rtc_get_alarm(struct tm *time)
+{
+    /* Convert from seconds since the epoch */
+    /* Note: Cast tells the compiler to discard volatile */
+    gmtime_r((time_t *)&rtc_state.alarm, time);
+
+    return 0;
+}
+
+void rtc_clear_alarm(void)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2B);
+
+    /* Interrupt safe order of assignment */
+    rtc_state.alarm_cb = NULL;
+    rtc_state.alarm_arg = NULL;
+}
+
+void rtc_poweron(void)
+{
+    rtt_poweron();
+}
+
+void rtc_poweroff(void)
+{
+    rtt_poweroff();
+}
+
+void atmega_rtc_incr(void)
+{
+    rtc_state.time += 8;
+
+    /* Enable alarm only if it will trigger in < 8 seconds */
+    if ((rtc_state.alarm & 0xFFFFFFF8) == (rtc_state.time & 0xFFFFFFF8)) {
+        if (rtc_state.alarm <= rtc_state.time) {
+            /* Prevent alarm offset if time is too soon */
+            rtc_state.alarm_cb(rtc_state.alarm_arg);
+        }
+        else {
+            TIMSK2 |= (1 << OCIE2B);
+        }
+    }
+}
+
+void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                   | (1 << TCR2AUB) | (1 << TCR2BUB))) ;
+}
+
+ISR(TIMER2_COMPB_vect) {
+    __enter_isr();
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2B);
+
+    if (rtc_state.alarm_cb != NULL) {
+        rtc_state.alarm_cb(rtc_state.alarm_arg);
+    }
+
+    if (sched_context_switch_request) {
```

Comment:
Just keeping in mind that this has to be changed when rebasing with https://github.com/RIOT-OS/RIOT/pull/8988 merged